File: packages/sdk/langs/node/src/helpers/format.ts
/**
 * Format helper methods for the Node SDK.
 *
 * These are hand-written convenience wrappers that call the canonical
 * `format.apply` operation with pre-filled inline styles. They are NOT generated
 * from the contract and will not be overwritten by `pnpm run generate:all`.
 *
 * Usage:
 * ```ts
 * import { createSuperDocClient } from 'superdoc';
 * import { formatBold, formatItalic } from 'superdoc/helpers/format';
 *
 * const client = createSuperDocClient();
 * await client.connect();
 *
 * // Canonical form:
 * await formatBold(client.doc, { target: { kind: 'text', blockId: 'p1', range: { start: 0, end: 5 } } });
 *
 * // Flat-flag shorthand (normalized before dispatch):
 * await formatBold(client.doc, { blockId: 'p1', start: 0, end: 5 });
 * ```
 */

import type { InvokeOptions, OperationSpec } from '../runtime/transport-common.js';

/**
 * Minimal operation spec for `format.apply`. Used to invoke the canonical
 * operation through the runtime without depending on generated code.
 *
 * Only canonical params are listed here. Flat-flag shortcuts (blockId,
 * start, end) are accepted via FormatHelperParams and normalized into
 * a `target` object before invoke.
 */
const FORMAT_APPLY_SPEC: OperationSpec = {
  operationId: 'doc.format.apply',
  commandTokens: ['format', 'apply'],
  params: [
    { name: 'doc', kind: 'doc', type: 'string' },
    { name: 'sessionId', kind: 'doc', flag: 'session', type: 'string' },
    { name: 'target', kind: 'jsonFlag', type: 'json' },
    { name: 'inline', kind: 'jsonFlag', type: 'json' },
    { name: 'dryRun', kind: 'flag', type: 'boolean' },
    { name: 'changeMode', kind: 'flag', type: 'string' },
    { name: 'expectedRevision', kind: 'flag', type: 'string' },
  ],
};

export interface FormatHelperParams {
  doc?: string;
  sessionId?: string;
  target?: { kind: 'text'; blockId: string; range: { start: number; end: number } };
  /** Flat-flag shorthand for target.blockId (normalized before dispatch). */
  blockId?: string;
  /** Flat-flag shorthand for target.range.start (normalized before dispatch). */
  start?: number;
  /** Flat-flag shorthand for target.range.end (normalized before dispatch). */
  end?: number;
  dryRun?: boolean;
  changeMode?: 'direct' | 'tracked';
  expectedRevision?: string;
}

/**
 * Generic invoke function that works with the SuperDocRuntime.
 * The doc API proxy created by `createDocApi(runtime)` exposes generated methods,
 * but helpers call the runtime directly for forward-compatibility.
 */
type RuntimeInvokeFn = <T = unknown>(
  operation: OperationSpec,
  params?: Record<string, unknown>,
  options?: InvokeOptions,
) => Promise<T>;

/**
 * Normalizes flat-flag shorthand params (blockId, start, end) into a
 * canonical `target` object. If `target` is already provided, flat flags
 * are left untouched (the caller provided the canonical form directly).
 */
function normalizeFormatParams(params: FormatHelperParams): Record<string, unknown> {
  const { blockId, start, end, target, ...rest } = params;
  if (blockId !== undefined && target === undefined) {
    return {
      ...rest,
      target: { kind: 'text', blockId, range: { start: start ?? 0, end: end ?? 0 } },
    };
  }
  return params as Record<string, unknown>;
}

function mergeInlineStyles(params: FormatHelperParams, inline: Record<string, boolean>): Record<string, unknown> {
  return { ...normalizeFormatParams(params), inline };
}

/**
 * Apply bold formatting to a text range.
 *
 * Equivalent to `format.apply` with `inline: { bold: true }`.
 */
export function formatBold(invoke: RuntimeInvokeFn, params: FormatHelperParams = {}, options?: InvokeOptions) {
  return invoke(FORMAT_APPLY_SPEC, mergeInlineStyles(params, { bold: true }), options);
}

/**
 * Apply italic formatting to a text range.
 *
 * Equivalent to `format.apply` with `inline: { italic: true }`.
 */
export function formatItalic(invoke: RuntimeInvokeFn, params: FormatHelperParams = {}, options?: InvokeOptions) {
  return invoke(FORMAT_APPLY_SPEC, mergeInlineStyles(params, { italic: true }), options);
}

/**
 * Apply underline formatting to a text range.
 *
 * Equivalent to `format.apply` with `inline: { underline: true }`.
 */
export function formatUnderline(invoke: RuntimeInvokeFn, params: FormatHelperParams = {}, options?: InvokeOptions) {
  return invoke(FORMAT_APPLY_SPEC, mergeInlineStyles(params, { underline: true }), options);
}

/**
 * Apply strikethrough formatting to a text range.
 *
 * Equivalent to `format.apply` with `inline: { strike: true }`.
 */
export function formatStrikethrough(invoke: RuntimeInvokeFn, params: FormatHelperParams = {}, options?: InvokeOptions) {
  return invoke(FORMAT_APPLY_SPEC, mergeInlineStyles(params, { strike: true }), options);
}

File: packages/sdk/langs/python/superdoc/helpers/format.py
"""
Format helper functions for the SuperDoc Python SDK.

These are hand-written convenience wrappers that call the canonical
``format.apply`` operation with pre-filled inline styles.  They are NOT generated
from the contract and will not be overwritten by ``pnpm run generate:all``.

Usage::

    from superdoc import SuperDocClient
    from superdoc.helpers import format_bold, format_italic

    client = SuperDocClient()
    client.connect()

    # Canonical form:
    result = client.doc.format_apply(
        target={"kind": "text", "blockId": "p1", "range": {"start": 0, "end": 5}},
        inline={"bold": True},
    )

    # Flat-flag shorthand (normalized before dispatch):
    result = format_bold(client.doc, block_id="p1", start=0, end=5)
"""

from __future__ import annotations

from typing import Any, Optional, Protocol


class FormatApplyCallable(Protocol):
    """Protocol matching the generated ``doc.format_apply`` method."""

    def __call__(self, **kwargs: Any) -> Any: ...


class DocApi(Protocol):
    """Minimal protocol for the doc API object returned by the generated client."""

    format_apply: FormatApplyCallable


def _normalize_target(
    target: Optional[dict[str, Any]],
    block_id: Optional[str],
    start: Optional[int],
    end: Optional[int],
) -> Optional[dict[str, Any]]:
    """Convert flat flags (block_id, start, end) to a canonical target dict.

    If *target* is already provided, flat flags are ignored. If *block_id*
    is provided without *target*, a text-range target is constructed.
    """
    if target is not None:
        return target
    if block_id is not None:
        return {
            "kind": "text",
            "blockId": block_id,
            "range": {"start": start if start is not None else 0, "end": end if end is not None else 0},
        }
    return None


def _format_inline(
    doc: DocApi,
    inline: dict[str, bool],
    *,
    target: Optional[dict[str, Any]] = None,
    block_id: Optional[str] = None,
    start: Optional[int] = None,
    end: Optional[int] = None,
    dry_run: Optional[bool] = None,
    change_mode: Optional[str] = None,
    expected_revision: Optional[str] = None,
    **extra: Any,
) -> Any:
    """Internal dispatch -- merges ``inline`` and forwards to ``format.apply``.

    Flat-flag shortcuts (``block_id``, ``start``, ``end``) are normalized
    into a canonical ``target`` dict before calling the API.
    """
    kwargs: dict[str, Any] = {"inline": inline}

    resolved_target = _normalize_target(target, block_id, start, end)
    if resolved_target is not None:
        kwargs["target"] = resolved_target

    if dry_run is not None:
        kwargs["dry_run"] = dry_run
    if change_mode is not None:
        kwargs["change_mode"] = change_mode
    if expected_revision is not None:
        kwargs["expected_revision"] = expected_revision
    if "inline" in extra:
        raise TypeError("Cannot pass 'inline' directly; it is set by the format helper.")
    kwargs.update(extra)
    return doc.format_apply(**kwargs)


def format_bold(doc: DocApi, **kwargs: Any) -> Any:
    """Apply bold formatting.  Equivalent to ``format.apply(inline={"bold": True})``."""
    return _format_inline(doc, {"bold": True}, **kwargs)


def format_italic(doc: DocApi, **kwargs: Any) -> Any:
    """Apply italic formatting.  Equivalent to ``format.apply(inline={"italic": True})``."""
    return _format_inline(doc, {"italic": True}, **kwargs)


def format_underline(doc: DocApi, **kwargs: Any) -> Any:
    """Apply underline formatting.  Equivalent to ``format.apply(inline={"underline": True})``."""
    return _format_inline(doc, {"underline": True}, **kwargs)


def format_strikethrough(doc: DocApi, **kwargs: Any) -> Any:
    """Apply strikethrough formatting.  Equivalent to ``format.apply(inline={"strike": True})``."""
    return _format_inline(doc, {"strike": True}, **kwargs)
